### PR TITLE
Support host side rate limit configuration

### PR DESCRIPTION
**Pull request by Junchao-Mellanox**
**HLD:**
https://github.com/sonic-net/SONiC/pull/1049

**What I did?**

Add a SyslogCfg to handle syslog rate limit configuration change

### CODE REVIEW

File: scripts/hostcfgd [hunk 1]
```diff
@@ -1510,6 +1584,12 @@ class HostConfigDaemon:
         self.passwcfg.load(passwh)
         self.devmetacfg.load(dev_meta)
         self.mgmtifacecfg.load(mgmt_ifc, mgmt_vrf)
+        self.syslogcfg.load(syslog)
+
+        dev_meta = self.config_db.get_table('DEVICE_METADATA')
+        if 'localhost' in dev_meta:
+            if 'hostname' in dev_meta['localhost']:
+                self.hostname_cache = dev_meta['localhost']['hostname']
```
Comment (mkeda) [inline scripts/hostcfgd:1592]: Is this used anywhere?

File: scripts/hostcfgd [hunk 2]
```diff
@@ -1440,6 +1441,75 @@ class MgmtIfaceCfg(object):
...
+            data (dict): CONFIG DB data: {<field_name>: <field_value>}
+        """
+        new_interval = '0'
+        new_burst = '0'
+        if data:
+            new_interval = data.get(self.SYSLOG_RATE_LIMIT_INTERVAL, '0')
+            new_burst = data.get(self.SYSLOG_RATE_LIMIT_BURST, '0')
+
+        if new_interval == self.current_interval and new_burst == self.current_burst:
+            syslog.syslog(syslog.LOG_INFO, 'Syslog rate limit configuration does not change, ignore it')
```
Comment (mkeda) [inline scripts/hostcfgd:1472]: I'd recommend to drop this info log.  In our environment, we already have a lot of logs.  :)

**Review by mkeda — CHANGES_REQUESTED** [review]
Thank you for working to add this feature!

File: scripts/hostcfgd [hunk 3]
```diff
@@ -1440,6 +1441,75 @@ class MgmtIfaceCfg(object):
...
+            return
+
+        syslog.syslog(syslog.LOG_INFO, f'Configure syslog rate limit interval={new_interval}, burst={new_burst}')
+
+        run_cmd('systemctl reset-failed rsyslog-config rsyslog')
+        run_cmd('systemctl restart rsyslog-config')
```
[the remark this answers is not in the document]
Comment (Junchao-Mellanox) [reply scripts/hostcfgd:1478]: No, rsyslog-config will restart rsyslog service

File: scripts/hostcfgd [hunk 4]
```diff
@@ -1440,6 +1441,75 @@ class MgmtIfaceCfg(object):
...
+        if self.HOST_KEY in data:
+            self.syslog_update(data[self.HOST_KEY])
+
+    def parse_syslog_conf(self):
+        """Passe existing syslog conf and extract config values
+
+        Returns:
+            tuple: interval,burst,target_ip
+        """
+        interval = '0'
```
[the remark this answers is not in the document]
Comment (Junchao-Mellanox) [reply scripts/hostcfgd:1492]: For host side, the default behavior is no rate-limiting. So, '0' is the default value here.

File: scripts/hostcfgd (hunk 1 above)
Comment (Junchao-Mellanox) [reply scripts/hostcfgd:1592]: Will remove

File: scripts/hostcfgd [hunk 5]
```diff
@@ -1440,6 +1441,75 @@ class MgmtIfaceCfg(object):
...
+    def syslog_update(self, data):
+        """Update syslog related configuration
+
+        Args:
+            data (dict): CONFIG DB data: {<field_name>: <field_value>}
+        """
+        new_interval = '0'
+        new_burst = '0'
+        if data:
+            new_interval = data.get(self.SYSLOG_RATE_LIMIT_INTERVAL, '0')
```
[the remark this answers is not in the document]
Comment (Junchao-Mellanox) [reply scripts/hostcfgd:1468]: For host side, no rate limiting is the default behavior. So, '0' is the expected default value.

File: scripts/hostcfgd [hunk 6]
```diff
@@ -1440,6 +1441,75 @@ class MgmtIfaceCfg(object):
...
+            new_burst = data.get(self.SYSLOG_RATE_LIMIT_BURST, '0')
+
+        if new_interval == self.current_interval and new_burst == self.current_burst:
+            syslog.syslog(syslog.LOG_INFO, 'Syslog rate limit configuration does not change, ignore it')
+            return
+
+        syslog.syslog(syslog.LOG_INFO, f'Configure syslog rate limit interval={new_interval}, burst={new_burst}')
+
+        run_cmd('systemctl reset-failed rsyslog-config rsyslog')
```
[the remark this answers is not in the document]
Comment (Junchao-Mellanox) [reply scripts/hostcfgd:1477]: Will let run_cmd raise exception. Once exception is caught, current_interval and current_burst shall not be updated.

**Comment by Junchao-Mellanox** [comment]
Hi @mkeda, I have fixed/replied all comments. Would you please review the new changes?

**Comment by zhangyanzhao** [comment]
@mkeda  please help to review and approve the PR if you are ok. Thanks.

**Review by mkeda — APPROVED** [review]

**Comment by saiarcot895** [comment]
Why is the executable bit being unset for `scripts/hostcfgd`?

File: scripts/hostcfgd [hunk 7]
```diff
@@ -1465,6 +1466,77 @@ class MgmtIfaceCfg(object):
...
+    def __init__(self):
+        self.current_interval, self.current_burst = self.parse_syslog_conf()
+
+    def syslog_update(self, data):
+        """Update syslog related configuration
+
+        Args:
+            data (dict): CONFIG DB data: {<field_name>: <field_value>}
+        """
+        new_interval = '0'
```
Comment (saiarcot895) [inline scripts/hostcfgd:1490]: Any particular reason to have these as strings instead of integers?

**Comment by Junchao-Mellanox** [comment]
> Why is the executable bit being unset for `scripts/hostcfgd`?

Will fix.

File: scripts/hostcfgd (hunk 7 above)
Comment (Junchao-Mellanox) [reply scripts/hostcfgd:1490]: Because redis pass string to hostcfgd, hostcfgd need compare redis value with existing value

**Comment by Junchao-Mellanox** [comment]
Hi @saiarcot895, I have fixed the file permission, could you please check?

**Review by saiarcot895 — APPROVED** [review]

**Pull request merged**